### PR TITLE
Introduce Finetuning Harness for In-Situ Reinforcement Learning of Agentic Workflows

### PR DESCRIPTION
**Pull request by dnandakumar-nv**
## Summary

This PR introduces a comprehensive **finetuning harness** to the NeMo Agent Toolkit, enabling in-situ reinforcement learning (RL) of agentic LLM workflows. The harness is designed with a decoupled architecture that separates training logic from backends, allowing users to train agents with any RL framework while collecting trajectories workflow evalutations.

### Key Features

- **Three-component architecture**: `TrajectoryBuilder`, `TrainerAdapter`, and `Trainer` interfaces that can be implemented independently
- **Backend-agnostic design**: Train with OpenPipe ART, NeMo Aligner, or custom implementations
- **In-situ training**: Collect trajectories during actual workflow execution, not from offline datasets
- **Curriculum learning**: Progressively introduce harder examples during training
- **GRPO support**: Group trajectories by example for Group Relative Policy Optimization

### 1. Core Finetuning Harness (`src/nat/finetuning/`)

**New Interfaces:**

| Interface | Purpose |
|-----------|---------|
| `TrajectoryBuilder` | Runs evaluations, collects episodes, computes rewards, groups trajectories |
| `TrainerAdapter` | Bridges NAT to external training backends, handles job submission and monitoring |
| `Trainer` | Orchestrates the complete finetuning loop across epochs |

**Data Models (`src/nat/data_models/finetuning.py`):**

- `Trajectory`: Episode sequence with reward and metadata
- `EpisodeItem`: Individual conversation turns with role, content, and logprobs
- `TrajectoryCollection`: Grouped trajectories for GRPO optimization
- `FinetuneConfig`: Complete finetuning configuration
- `CurriculumLearningConfig`: Progressive difficulty settings
- `TrainingJobRef` / `TrainingJobStatus`: Job tracking primitives

**Runtime (`src/nat/finetuning/finetuning_runtime.py`):**

- Async finetuning execution
- Component initialization and binding
- Epoch orchestration with validation support

**CLI Command (`src/nat/cli/commands/finetune.py`):**

```bash
nat finetune --config_file=config.yml \
    --validation_dataset=val.json \
    --validation_interval=5
```

### 2. OpenPipe ART Plugin (`packages/nvidia_nat_openpipe_art/`)

A complete implementation of the finetuning interfaces for [OpenPipe ART](https://art.openpipe.ai/):

| Component | Description |
|-----------|-------------|
| `ARTTrajectoryBuilder` | Parallel evaluation runs, logprob extraction, reward computation |
| `ARTTrainerAdapter` | ART client management, trajectory format conversion, async job submission |
| `ARTTrainer` | Epoch orchestration, curriculum learning, validation, progress visualization |

**Configuration types:**
- `openpipe_art_traj_builder`
- `openpipe_art_trainer_adapter`
- `openpipe_art_trainer`

### 3. Tic-Tac-Toe RL Example (`examples/finetuning/rl_with_openpipe_art/`)

A complete working example demonstrating GRPO training:

- **Task**: LLM plays Tic-Tac-Toe against a random opponent
- **Reward function**: Alpha-beta pruning based position evaluation (continuous, dense rewards)
- **Workflow**: Full game loop with intermediate step tracking
- **Evaluators**: Win percentage and step-value computation
- **Configs**: Pre-training, training, and post-training evaluation

The example includes a sophisticated reward function that:
- Uses game-theoretic minimax with alpha-beta pruning to identify forced wins/losses
- Provides continuous rewards in [0, 1] for non-terminal positions
- Applies temporal discounting to weight earlier moves more heavily
- Adds bonus rewards for winning positions

### 4. Documentation (`docs/source/reference/finetuning/`)

Comprehensive documentation split into focused guides:

| Document | Content |
|----------|---------|
| `index.md` | Overview, architecture diagram, quick start |
| `concepts.md` | RL fundamentals, trajectories, curriculum learning, credit assignment, GRPO/PPO/DPO |
| `extending.md` | Step-by-step guide to implementing custom components |
| `rl_with_openpipe.md` | OpenPipe ART integration details, configuration reference |

### 5. Tests

- Unit tests for finetuning data models
- Integration tests for trajectory building
- End-to-end tests for the ART integration
- Example workflow tests

## Architecture

```
┌─────────────────────────────────────────────────────────────────────────┐
│                              Trainer                                     │
│  (Orchestrates the finetuning loop across epochs)                       │
│                                                                         │
│  ┌───────────────────────┐         ┌───────────────────────────┐       │
│  │  TrajectoryBuilder    │         │    TrainerAdapter         │       │
│  │                       │         │                           │       │
│  │  - Runs evaluations   │ ──────► │  - Validates trajectories │       │
│  │  - Collects episodes  │         │  - Submits to backend     │       │
│  │  - Computes rewards   │         │  - Monitors training      │       │
│  │  - Groups trajectories│         │  - Reports status         │       │
│  └───────────────────────┘         └───────────────────────────┘       │
└─────────────────────────────────────────────────────────────────────────┘
                                         │
                                         ▼
                            ┌─────────────────────────┐
                            │   Remote Training       │
                            │      Backend            │
                            │  (OpenPipe ART, etc.)   │
                            └─────────────────────────┘
```

## Configuration Example

```yaml
llms:
  training_llm:
    _type: openai
    model_name: Qwen/Qwen2.5-3B-Instruct
    base_url: http://localhost:8000/v1

workflow:
  _type: my_workflow
  llm: training_llm

eval:
  general:
    max_concurrency: 16
    dataset:
      _type: json
      file_path: data/training.json
  evaluators:
    reward:
      _type: my_evaluator

trajectory_builders:
  builder:
    _type: openpipe_art_traj_builder
    num_generations: 2

trainer_adapters:
  adapter:
    _type: openpipe_art_trainer_adapter
    backend:
      ip: localhost
      port: 7623
      base_model: Qwen/Qwen2.5-3B-Instruct
    training:
      learning_rate: 1e-6

trainers:
  trainer:
    _type: openpipe_art_trainer

finetuning:
  enabled: true
  trainer: trainer
  trajectory_builder: builder
  trainer_adapter: adapter
  reward_function:
    name: reward
  num_epochs: 20
  curriculum_learning:
    enabled: true
    initial_percentile: 0.3
    expansion_interval: 5
```

### Why Decoupled Architecture?

1. **Backend flexibility**: Organizations may have existing training infrastructure (NeMo Customizer, custom GRPO implementations) they want to use
2. **Algorithm agnosticism**: Support GRPO, PPO, DPO, or SFT without code changes
3. **Testability**: Each component can be tested independently with mocks
4. **Future-proofing**: New training backends can be added without modifying core logic

### Why In-Situ Training?

1. **Real data distribution**: Training data reflects actual use cases and edge conditions
2. **Continuous improvement**: Models improve while the system remains operational
3. **Tighter feedback loops**: Hours instead of weeks between training iterations
4. **Adaptive**: Naturally handles distribution shift

## Breaking Changes

None. This is a new feature addition.

## Migration Guide

N/A - New feature.

## Testing

```bash
# Run finetuning harness tests
pytest tests/nat/finetuning/

# Run OpenPipe ART plugin tests
pytest packages/nvidia_nat_openpipe_art/tests/

# Run example tests
pytest examples/finetuning/rl_with_openpipe_art/tests/
```

### Running the Example

1. **Start the ART server** (requires GPU):
   ```bash
   source ~/.venvs/art-env/bin/activate
   uv run art --host 0.0.0.0 --port 7623
   ```

2. **Run pre-training baseline**:
   ```bash
   nat eval --config_file=examples/finetuning/rl_with_openpipe_art/src/rl_with_openpipe_art/configs/config_pre_train.yml
   ```

3. **Run finetuning**:
   ```bash
   nat finetune --config_file=examples/finetuning/rl_with_openpipe_art/src/rl_with_openpipe_art/configs/config.yml
   ```

4. **Run post-training evaluation**:
   ```bash
   nat eval --config_file=examples/finetuning/rl_with_openpipe_art/src/rl_with_openpipe_art/configs/config_post_train.yml
   ```

## Checklist

- [x] Core finetuning interfaces implemented
- [x] Finetuning runtime and CLI command
- [x] Data models for trajectories, episodes, and configurations
- [x] OpenPipe ART plugin package
- [x] Tic-Tac-Toe RL example with alpha-beta reward function
- [x] Comprehensive documentation (concepts, extending, ART integration)
- [x] Unit and integration tests
- [x] Example README with step-by-step instructions

## Future Work in other PRs

- Additional training backends (NeMo Customizer)
- DPO/SFT trajectory builder variants
- Checkpoint export utilities

**Reviewers**: Please pay particular attention to:
1. Interface design in `src/nat/finetuning/interfaces/`
2. Data model structure in `src/nat/data_models/finetuning.py`
3. Documentation completeness in `docs/source/reference/finetuning/`

## Summary by CodeRabbit

* **New Features**
  * Experimental finetuning workflow: finetune CLI, OpenPipe ART backend support, new finetuning APIs, and builder/workflow integrations.
  * Example: end-to-end Tic‑Tac‑Toe RL example with configs, evaluator, agents, and runnable finetuning configs.

* **Documentation**
  * New comprehensive finetuning docs (concepts, extending, OpenPipe integration, CLI).

* **Packaging & Examples**
  * Optional OpenPipe ART package and workspace wiring; example added to workspace.

* **Tests**
  * Extensive unit tests for finetuning components, adapters, builders, and parsers.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/nvidia_nat_openpipe_art/LICENSE-3rd-party.txt
Comment (willkill07) [inline packages/nvidia_nat_openpipe_art/LICENSE-3rd-party.txt:1]: Should be a symlink to parent package

File: packages/nvidia_nat_openpipe_art/LICENSE.md
Comment (willkill07) [inline packages/nvidia_nat_openpipe_art/LICENSE.md:1]: Should be a symlink to parent package

File: packages/nvidia_nat_openpipe_art/pyproject.toml
Comment (dagardner-nv) [inline packages/nvidia_nat_openpipe_art/pyproject.toml:1]: Don't forget to add this new package as a dependency in `packages/nvidia_nat_all/pyproject.toml`

**Review by dagardner-nv — APPROVED** [review]

**Review by willkill07 — APPROVED** [review]
Minor nits that need to be corrected before merge, but LGTM overall.
